deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/constraint/BaseConstraint.java

```diff
@@ -0,0 +1,64 @@
+package org.deeplearning4j.nn.conf.constraint;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import org.apache.commons.lang3.ArrayUtils;
+import org.deeplearning4j.nn.api.Layer;
+import org.deeplearning4j.nn.api.ParamInitializer;
+import org.deeplearning4j.nn.api.layers.LayerConstraint;
+import org.nd4j.linalg.api.ndarray.INDArray;
+
+import java.util.Map;
+
+
+@AllArgsConstructor
+@EqualsAndHashCode
+@Data
+public abstract class BaseConstraint implements LayerConstraint {
+    public static final double DEFAULT_EPSILON = 1e-6;
+
+    protected boolean applyToWeights;
+    protected boolean applyToBiases;
+    protected double epsilon = 1e-6;
+    protected int[] dimensions;
+
+    protected BaseConstraint(){
+        //No arg for json ser/de
+    }
+
+    protected BaseConstraint(boolean applyToWeights, boolean applyToBiases, int... dimensions){
+        this(applyToWeights, applyToBiases, DEFAULT_EPSILON, dimensions);
+    }
+
+
+    @Override
+    public void applyConstraint(Layer layer, int iteration, int epoch) {
+        Map<String,INDArray> paramTable = layer.paramTable();
+        if(paramTable == null || paramTable.isEmpty() ){
+            return;
+        }
+
+        ParamInitializer i = layer.conf().getLayer().initializer();
+        for(Map.Entry<String,INDArray> e : paramTable.entrySet()){
+            if(applyToWeights && i.isWeightParam(e.getKey()) || applyToBiases && i.isBiasParam(e.getKey())){
+                apply(e.getValue(), i.isBiasParam(e.getKey()));
```

Comment:
maybe we can replace `i.isBiasParam(e.getKey()` here simply with `true`?! It took me a while to realise that it doesn't matter what to use here, but this seems to indicate it has specifically to to with bias, which it doesn't

Reply:
No - some parameters are neither weights or biases... Consider for example batch normalization statistics, which are technically parameters in DL4J, but neither weights or biases.